Engine/DataFeeds/Enumerators/CorporateEventBaseEnumerator.cs

```diff
@@ -0,0 +1,188 @@
+        /// </summary>
+        protected BaseData PreviousUnderlyingData;
+
+        /// <summary>
+        ///The Subscription data config used by this enumerator
+        /// </summary>
+        protected readonly SubscriptionDataConfig SubscriptionDataConfig;
+
+        /// <summary>
+        /// Creates a new instance
+        /// </summary>
+        /// <param name="enumerator">Underlying enumerator</param>
+        /// <param name="config">The <see cref="SubscriptionDataConfig"/></param>
+        /// <param name="tradableDayNotifier">Tradable dates provider</param>
+        /// <param name="includeAuxiliaryData">True to emit auxiliary data</param>
+        public CorporateEventBaseEnumerator(
+            IEnumerator<BaseData> enumerator,
+            SubscriptionDataConfig config,
+            ITradableDatesNotifier tradableDayNotifier,
+            bool includeAuxiliaryData)
+        {
+            SubscriptionDataConfig = config;
+            _auxiliaryData = new Queue<BaseData>();
+            _enumerator = enumerator;
+            tradableDayNotifier.NewTradableDate += (sender, dateTime) =>
+            {
+                // Call implementation
+                var newEvent = CheckNewEvent(dateTime);
+                if (newEvent != null && includeAuxiliaryData)
+                {
+                    _auxiliaryData.Enqueue(newEvent);
+                }
+            };
+        }
+
+        /// <summary>
+        /// Called each time there is a new tradable day
+        /// </summary>
+        /// <param name="date">The new tradable day value</param>
+        /// <returns>New corporate event, else Null</returns>
+        protected abstract BaseData CheckNewEvent(DateTime date);
+
+        /// <summary>
+        /// Advances the enumerator to the next element.
+        /// Will merge chronologically new data produced by this enumerator
+        /// with underlying enumerator.
+        /// </summary>
+        /// <returns>
+        /// True if the enumerator was successfully advanced to the next element;
+        /// </returns>
+        public virtual bool MoveNext()
+        {
+            // save previous value
+            PreviousUnderlyingData = _enumerator.Current;
+
+            // if we emitted data we don't have to turn underlying enumerator
+            // because we already did in previous step
+            if (!_weEmittedData
+                && !_underlyingExhausted
+                && !_enumerator.MoveNext())
+            {
+                _underlyingExhausted = true;
+            }
+
+            _weEmittedData = false;
+            if (_auxiliaryData.Any()
+                && (_enumerator.Current == null // if underlying is null but we have data, emit
+                    || _auxiliaryData.Peek().EndTime < _enumerator.Current.EndTime
+                    || _underlyingExhausted))
+            {
+                _weEmittedData = true;
+                Current = _auxiliaryData.Dequeue();
```

Comment:
I think a main goal of this dedicated enumerator was removing the need for an internal aux queue. If you like, we can throw ideas around, but this implementation seems to suffer from the same disadvantages that the original implementation did, albeit, defined in a separate class.

Reply:
Using the `SynchronizingEnumerator` I'll be able to lose almost all this ugly logic, I would still need a `Queue` to hold the generated events. Let me know what you think of the new approach 👍